misc/multistream-select/src/negotiated.rs

```diff
@@ -125,11 +129,12 @@ impl<TInner> Negotiated<TInner> {
                         }
                     };
 
-                    if let Message::Header(v) = &msg {
-                        if *v == version {
-                            *this.state = State::Expecting { io, protocol, version };
+                    match (&msg, version) {
+                        (Message::Header(Version::V1), Some(Version::V1)) => {
```

Comment:
The fact that we assume that `version` can never be `V1Lazy` looks a bit like a hack to me.

In an ideal world, we would differentiate the `Version` enum (containing `V1` and `V1Lazy`) from the `VersionHandshake` (containing only `V1`), no?

Reply:
> The fact that we assume that version can never be V1Lazy looks a bit like a hack to me.

By definition of these versions, they're now indistinguishable on the wire and it is always expected to receive `V1`. So receiving `V1Lazy` is considered an error (in the decoder really).

> In an ideal world, we would differentiate the Version enum (containing V1 and V1Lazy) from the VersionHandshake (containing only V1), no?

Since `V1Lazy` is now really the same negotiation protocol as `V1`, only differing subtly in the behaviour of the dialer, it may be cleaner to separate this "lazy flush yes/no" setting from the `Version`, making it a separate configuration option for the dialer only, since `V1Lazy` used for a listener now has no effect at all. However, for transport and substream upgrades we currently only have this choice of `Version` without a concept of separate parameters for the roles of dialer and/or listener. Maybe `Version::V1 { dialer_lazy_flush: bool }` would already be an improvement, though be another API change. It would always be decoded from the wire with `dialer_lazy_flush: false` since that option is local. Should I give that a try?  So due to the fact that this would require larger API changes, together with my hope that `V1Lazy` will eventually disappear altogether, merging its behaviour into `V1`, led me to the current approach, but I'm open to alternatives.

Reply:
> It would always be decoded from the wire with dialer_lazy_flush: false since that option is local.

What I had in mind is that you wouldn't decode a `Version`, but a new enum called for example `VersionHandshake` that would only contain `V1` without any parameter.

I don't know if that makes sense, and I'm also not strongly opinionated. I'm fine with these changes.

Reply:
> What I had in mind is that you wouldn't decode a Version, but a new enum called for example VersionHandshake that would only contain V1 without any parameter.

I see, separating these two kinds of versions. I will give that a try.

Reply:
My proposal is https://github.com/libp2p/rust-libp2p/pull/1855/commits/70f041b8a79fb0f6e21048577e71188345dc2c25.